File: python/src/alp/crypto.py
from __future__ import annotations

import base64
import hashlib
import secrets
from copy import deepcopy
from datetime import UTC, datetime, timedelta
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from .canonical import canonical_json_bytes
from .errors import ALPAuthError


def utc_now() -> datetime:
    return datetime.now(UTC)


def utc_now_iso() -> str:
    return utc_now().isoformat().replace("+00:00", "Z")


def parse_iso8601(value: str) -> datetime:
    if value.endswith("Z"):
        value = value[:-1] + "+00:00"
    return datetime.fromisoformat(value).astimezone(UTC)


def b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def b64url_decode(data: str) -> bytes:
    padding = "=" * (-len(data) % 4)
    return base64.urlsafe_b64decode(data + padding)


def generate_nonce() -> str:
    return b64url_encode(secrets.token_bytes(18))


def generate_keypair() -> tuple[str, str]:
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()
    private_raw = private_key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    public_raw = public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    return b64url_encode(private_raw), b64url_encode(public_raw)


def load_private_key(key_material: str | bytes) -> Ed25519PrivateKey:
    if isinstance(key_material, bytes):
        raw = key_material
    else:
        raw = b64url_decode(key_material)
    return Ed25519PrivateKey.from_private_bytes(raw)


def load_public_key(key_material: str | bytes) -> Ed25519PublicKey:
    if isinstance(key_material, bytes):
        raw = key_material
    else:
        raw = b64url_decode(key_material)
    return Ed25519PublicKey.from_public_bytes(raw)


def sha256_hex(value: Any) -> str:
    return hashlib.sha256(canonical_json_bytes(value)).hexdigest()


def payload_without_signature(payload: dict[str, Any]) -> dict[str, Any]:
    clean = deepcopy(payload)
    auth = clean.get("auth")
    if isinstance(auth, dict):
        auth["signature"] = ""
    return clean


def sign_protocol_object(payload: dict[str, Any], private_key_material: str | bytes, *, ttl_seconds: int = 300) -> dict[str, Any]:
    signed = deepcopy(payload)
    auth = signed.setdefault("auth", {})
    auth.setdefault("alg", "Ed25519")
    if not auth.get("nonce"):
        auth["nonce"] = generate_nonce()
    auth["issued_at"] = utc_now_iso()
    auth["expires_at"] = (utc_now() + timedelta(seconds=ttl_seconds)).isoformat().replace("+00:00", "Z")
    if not auth.get("key_id"):
        raise ALPAuthError("auth.key_id is required before signing")
    unsigned = payload_without_signature(signed)
    signature = load_private_key(private_key_material).sign(canonical_json_bytes(unsigned))
    auth["signature"] = b64url_encode(signature)
    return signed


def verify_protocol_object(payload: dict[str, Any], public_key_material: str | bytes, *, now: datetime | None = None) -> None:
    auth = payload.get("auth") or {}
    signature = auth.get("signature")
    if not signature:
        raise ALPAuthError("missing signature")
    issued_at = auth.get("issued_at")
    expires_at = auth.get("expires_at")
    if not issued_at or not expires_at:
        raise ALPAuthError("missing signature timestamps")
    reference_time = now or utc_now()
    if parse_iso8601(issued_at) > reference_time + timedelta(minutes=5):
        raise ALPAuthError("signature issued_at is in the future")
    if parse_iso8601(expires_at) < reference_time:
        raise ALPAuthError("signature has expired")
    try:
        load_public_key(public_key_material).verify(b64url_decode(signature), canonical_json_bytes(payload_without_signature(payload)))
    except InvalidSignature as exc:
        raise ALPAuthError("invalid signature") from exc


def sign_status_request(path: str, issuer: str, key_id: str, private_key_material: str | bytes, *, ttl_seconds: int = 300) -> dict[str, str]:
    issued_at = utc_now_iso()
    expires_at = (utc_now() + timedelta(seconds=ttl_seconds)).isoformat().replace("+00:00", "Z")
    nonce = generate_nonce()
    message = f"GET\n{path}\n{issuer}\n{issued_at}\n{expires_at}\n{nonce}".encode("utf-8")
    signature = load_private_key(private_key_material).sign(message)
    return {
        "X-ALP-Issuer": issuer,
        "X-ALP-Key-Id": key_id,
        "X-ALP-Issued-At": issued_at,
        "X-ALP-Expires-At": expires_at,
        "X-ALP-Nonce": nonce,
        "X-ALP-Signature": b64url_encode(signature),
    }


def verify_status_headers(path: str, headers: dict[str, str], public_key_material: str | bytes, *, now: datetime | None = None) -> str:
    try:
        issuer = headers["X-ALP-Issuer"]
        issued_at = headers["X-ALP-Issued-At"]
        expires_at = headers["X-ALP-Expires-At"]
        nonce = headers["X-ALP-Nonce"]
        signature = headers["X-ALP-Signature"]
    except KeyError as exc:
        raise ALPAuthError(f"missing status header: {exc.args[0]}") from exc
    reference_time = now or utc_now()
    if parse_iso8601(expires_at) < reference_time:
        raise ALPAuthError("status request signature has expired")
    message = f"GET\n{path}\n{issuer}\n{issued_at}\n{expires_at}\n{nonce}".encode("utf-8")
    try:
        load_public_key(public_key_material).verify(b64url_decode(signature), message)
    except InvalidSignature as exc:
        raise ALPAuthError("invalid status request signature") from exc
    return issuer

File: typescript/src/crypto.ts
import { createHash } from "node:crypto";
import nacl from "tweetnacl";

import { canonicalJson } from "./canonical.js";
import { ALPAuthError } from "./errors.js";

export function utcNowIso(): string {
  return new Date().toISOString();
}

function b64urlEncode(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64url");
}

function b64urlDecode(value: string): Uint8Array {
  return new Uint8Array(Buffer.from(value, "base64url"));
}

export function generateNonce(): string {
  return b64urlEncode(nacl.randomBytes(18));
}

export function generateKeypair(): { privateKey: string; publicKey: string } {
  const pair = nacl.sign.keyPair();
  const seed = pair.secretKey.slice(0, 32);
  return {
    privateKey: b64urlEncode(seed),
    publicKey: b64urlEncode(pair.publicKey)
  };
}

function keyPairFromSeed(privateKey: string | Uint8Array) {
  const seed = typeof privateKey === "string" ? b64urlDecode(privateKey) : privateKey;
  return nacl.sign.keyPair.fromSeed(seed);
}

function payloadWithoutSignature(payload: Record<string, unknown>): Record<string, unknown> {
  const copy = structuredClone(payload);
  const auth = (copy.auth ?? {}) as Record<string, unknown>;
  auth.signature = "";
  copy.auth = auth;
  return copy;
}

export function sha256Hex(value: unknown): string {
  return createHash("sha256").update(canonicalJson(value)).digest("hex");
}

export function signProtocolObject<T extends Record<string, unknown>>(
  payload: T,
  privateKey: string | Uint8Array,
  ttlSeconds = 300
): T {
  const signed = structuredClone(payload);
  const auth = (((signed as Record<string, unknown>).auth as Record<string, unknown> | undefined) ?? {}) as Record<string, unknown>;
  auth.alg = "Ed25519";
  auth.issued_at = utcNowIso();
  auth.expires_at = new Date(Date.now() + ttlSeconds * 1000).toISOString();
  auth.nonce = typeof auth.nonce === "string" && auth.nonce.length > 0 ? auth.nonce : generateNonce();
  if (!auth.key_id || typeof auth.key_id !== "string") {
    throw new ALPAuthError("auth.key_id is required before signing");
  }
  (signed as Record<string, unknown>).auth = auth;
  const pair = keyPairFromSeed(privateKey);
  const signature = nacl.sign.detached(Buffer.from(canonicalJson(payloadWithoutSignature(signed))), pair.secretKey);
  auth.signature = b64urlEncode(signature);
  return signed;
}

export function verifyProtocolObject(payload: Record<string, unknown>, publicKey: string | Uint8Array): void {
  const auth = (payload.auth ?? {}) as Record<string, unknown>;
  const signature = auth.signature;
  if (typeof signature !== "string" || !signature) {
    throw new ALPAuthError("missing signature");
  }
  const expiresAt = auth.expires_at;
  if (typeof expiresAt !== "string" || new Date(expiresAt).getTime() < Date.now()) {
    throw new ALPAuthError("signature has expired");
  }
  const key = typeof publicKey === "string" ? b64urlDecode(publicKey) : publicKey;
  const ok = nacl.sign.detached.verify(
    Buffer.from(canonicalJson(payloadWithoutSignature(payload))),
    b64urlDecode(signature),
    key
  );
  if (!ok) {
    throw new ALPAuthError("invalid signature");
  }
}

export function signStatusRequest(
  path: string,
  issuer: string,
  keyId: string,
  privateKey: string | Uint8Array,
  ttlSeconds = 300
): Record<string, string> {
  const issuedAt = utcNowIso();
  const expiresAt = new Date(Date.now() + ttlSeconds * 1000).toISOString();
  const nonce = generateNonce();
  const message = Buffer.from(`GET\n${path}\n${issuer}\n${issuedAt}\n${expiresAt}\n${nonce}`);
  const pair = keyPairFromSeed(privateKey);
  const signature = nacl.sign.detached(message, pair.secretKey);
  return {
    "X-ALP-Issuer": issuer,
    "X-ALP-Key-Id": keyId,
    "X-ALP-Issued-At": issuedAt,
    "X-ALP-Expires-At": expiresAt,
    "X-ALP-Nonce": nonce,
    "X-ALP-Signature": b64urlEncode(signature)
  };
}

export function verifyStatusHeaders(
  path: string,
  headers: Record<string, string | string[] | undefined>,
  publicKey: string | Uint8Array
): string {
  const issuer = headers["x-alp-issuer"] as string | undefined;
  const issuedAt = headers["x-alp-issued-at"] as string | undefined;
  const expiresAt = headers["x-alp-expires-at"] as string | undefined;
  const nonce = headers["x-alp-nonce"] as string | undefined;
  const signature = headers["x-alp-signature"] as string | undefined;
  if (!issuer || !issuedAt || !expiresAt || !nonce || !signature) {
    throw new ALPAuthError("missing signed status headers");
  }
  if (new Date(expiresAt).getTime() < Date.now()) {
    throw new ALPAuthError("status signature expired");
  }
  const message = Buffer.from(`GET\n${path}\n${issuer}\n${issuedAt}\n${expiresAt}\n${nonce}`);
  const key = typeof publicKey === "string" ? b64urlDecode(publicKey) : publicKey;
  const ok = nacl.sign.detached.verify(message, b64urlDecode(signature), key);
  if (!ok) {
    throw new ALPAuthError("invalid status request signature");
  }
  return issuer;
}
